File: app/src/main/java/im/tox/antox/adapters/ChatMessagesAdapter.java
package im.tox.antox.adapters;

import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.database.Cursor;
import android.net.Uri;
import android.os.AsyncTask;
import android.os.Environment;
import android.support.v4.widget.ResourceCursorAdapter;
import android.text.ClipboardManager;
import android.util.Log;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;

import java.io.File;
import java.sql.Timestamp;
import java.util.HashSet;

import im.tox.antox.R;
import im.tox.antox.data.AntoxDB;
import im.tox.antox.tox.ToxSingleton;
import im.tox.antox.utils.BitmapManager;
import im.tox.antox.utils.ChatMessages;
import im.tox.antox.utils.Constants;
import im.tox.antox.utils.PrettyTimestamp;
import im.tox.antox.utils.Tuple;

public class ChatMessagesAdapter extends ResourceCursorAdapter {
    Context context;
    int layoutResourceId = R.layout.chat_message_row;
    private int density;
    private int paddingscale = 8;
    private ToxSingleton toxSingleton = ToxSingleton.getInstance();
    private Animation anim;
    private LayoutInflater mInflater;
    private HashSet<Integer> animatedIds;

    public ChatMessagesAdapter(Context context, Cursor c, HashSet<Integer> ids) {
        super(context, R.layout.chat_message_row, c, 0);
        this.context = context;
        this.anim = AnimationUtils.loadAnimation(this.context, R.anim.abc_slide_in_bottom);
        mInflater = (LayoutInflater) context.getSystemService(Context.LAYOUT_INFLATER_SERVICE);
        this.density = (int) context.getResources().getDisplayMetrics().density;
        this.animatedIds = ids;
    }


    @Override
     public View newView(Context context, Cursor cursor, ViewGroup parent) {
          return mInflater.inflate(this.layoutResourceId, parent, false);
     }
    @Override
    public void bindView(View view, final Context context, Cursor cursor) {
        final int id = cursor.getInt(0);
        final Timestamp time = Timestamp.valueOf(cursor.getString(1));
        final int message_id = cursor.getInt(2);
        final String k = cursor.getString(3);
        final String m = cursor.getString(4);
        final boolean received = cursor.getInt(5)>0;
        final boolean read = cursor.getInt(6)>0;
        final boolean sent = cursor.getInt(7)>0;
        final int size = cursor.getInt(8);
        final int type = cursor.getInt(9);


        ChatMessages msg = new ChatMessages(id, message_id, m, time, received, sent, size, type);
        ChatMessagesHolder holder = new ChatMessagesHolder();


        holder.message = (TextView) view.findViewById(R.id.message_text);
        holder.layout = (LinearLayout) view.findViewById(R.id.message_text_layout);
        holder.row = (LinearLayout) view.findViewById(R.id.message_row_layout);
        holder.background = (LinearLayout) view.findViewById(R.id.message_text_background);
        holder.time = (TextView) view.findViewById(R.id.message_text_date);
        holder.sent = (ImageView) view.findViewById(R.id.chat_row_sent);
        holder.received = (ImageView) view.findViewById(R.id.chat_row_received);
        holder.title = (TextView) view.findViewById(R.id.message_title);
        holder.progress = (ProgressBar) view.findViewById(R.id.file_transfer_progress);
        holder.imageMessage = (ImageView) view.findViewById(R.id.message_sent_photo);
        holder.imageMessageFrame = (FrameLayout) view.findViewById(R.id.message_sent_photo_frame);
        holder.progressText = (TextView) view.findViewById(R.id.file_transfer_progress_text);
        holder.padding = (View) view.findViewById(R.id.file_transfer_padding);
        holder.buttons = (LinearLayout) view.findViewById(R.id.file_buttons);
        holder.accept = (View) view.findViewById(R.id.file_accept_button);
        holder.reject = (View) view.findViewById(R.id.file_reject_button);

        holder.message.setTextSize(16);
        holder.message.setVisibility(View.GONE);
        holder.time.setVisibility(View.GONE);
        holder.sent.setVisibility(View.GONE);
        holder.received.setVisibility(View.GONE);
        holder.title.setVisibility(View.GONE);
        holder.progress.setVisibility(View.GONE);
        holder.imageMessage.setVisibility(View.GONE);
        holder.imageMessageFrame.setVisibility(View.GONE);
        holder.progressText.setVisibility(View.GONE);
        holder.padding.setVisibility(View.GONE);
        holder.buttons.setVisibility(View.GONE);
        switch(type) {
            case Constants.MESSAGE_TYPE_OWN:
                ownMessage(holder);
                holder.message.setText(msg.message);
                holder.message.setVisibility(View.VISIBLE);
                if (msg.sent) {
                    if (msg.received) {
                        holder.received.setVisibility(View.VISIBLE);
                    } else {
                        holder.sent.setVisibility(View.VISIBLE);
                    }
                }
                break;

            case Constants.MESSAGE_TYPE_FRIEND:
                friendMessage(holder);
                holder.message.setText(msg.message);
                holder.message.setVisibility(View.VISIBLE);
                break;

            case Constants.MESSAGE_TYPE_FILE_TRANSFER:
            case Constants.MESSAGE_TYPE_FILE_TRANSFER_FRIEND:
                toxSingleton.fileSizeMap.put(id,size);
                if (type == Constants.MESSAGE_TYPE_FILE_TRANSFER) {
                    ownMessage(holder);
                    String[] split = msg.message.split("/");
                    holder.message.setText(split[split.length - 1]);
                    holder.message.setVisibility(View.VISIBLE);
                } else {
                    friendMessage(holder);
                    holder.message.setText(msg.message);
                    holder.message.setVisibility(View.VISIBLE);
                }

                holder.title.setVisibility(View.VISIBLE);
                holder.title.setText(R.string.chat_file_transfer);

                if (msg.received) {
                    holder.progressText.setText("Finished");
                    holder.progressText.setVisibility(View.VISIBLE);
                } else {
                    if (msg.sent) {
                        if (msg.message_id != -1) {
                            holder.progress.setVisibility(View.VISIBLE);
                            holder.progress.setMax(msg.size);
                            holder.progress.setProgress(toxSingleton.getProgress(msg.id));
                            Tuple<Integer, Long> progress = toxSingleton.getProgressSinceXAgo(msg.id, 5000);
                            int bytesPerSecond;
                            if (progress != null) {
                                bytesPerSecond = (int) ((long) (progress.x * 1000) / progress.y);
                            } else {
                                bytesPerSecond = 0;
                            }
                            if (bytesPerSecond != 0) {
                                int secondsToComplete = msg.size / bytesPerSecond;
                                holder.progressText.setText(Integer.toString(bytesPerSecond/1024) + " KiB/s, " + secondsToComplete + " seconds left");
                            } else {
                                holder.progressText.setText(Integer.toString(bytesPerSecond / 1024) + " KiB/s");
                            }
                            holder.progressText.setVisibility(View.VISIBLE);
                        } else { //Filesending failed, it's sent, we no longer have a filenumber, but it hasn't been received
                            holder.progressText.setText("Failed");
                            holder.progressText.setVisibility(View.VISIBLE);
                        }
                    } else {
                        if (msg.message_id != -1) {
                            if (msg.isMine()) {
                                holder.progressText.setText("Sent filesending request");
                            } else {
                                holder.progressText.setText("");
                                holder.buttons.setVisibility(View.VISIBLE);
                                holder.accept.setOnClickListener(new View.OnClickListener() {
                                    @Override
                                    public void onClick(View view) {
                                        toxSingleton.acceptFile(k, message_id, context);
                                    }
                                });
                                holder.reject.setOnClickListener(new View.OnClickListener() {
                                    @Override
                                    public void onClick(View view) {
                                        toxSingleton.rejectFile(k, message_id, context);
                                    }
                                });
                            }
                        } else { //Filesending request not accepted, we no longer have a filenumber, but it hasn't been accepted
                            holder.progressText.setText("Rejected");
                        }
                        holder.progressText.setVisibility(View.VISIBLE);
                    }
                }

                if (msg.received || msg.isMine()) {
                    File f = null;
                    if (msg.message.contains("/")) {
                        f = new File(msg.message);
                    } else {
                        f = new File(Environment.getExternalStoragePublicDirectory(
                                Environment.DIRECTORY_DOWNLOADS), Constants.DOWNLOAD_DIRECTORY);
                        f = new File(f.getAbsolutePath() + "/" + msg.message);
                    }

                    if (f.exists()) {
                        final File file = f;
                        final String[] okFileExtensions =  new String[] {"jpg", "png", "gif","jpeg"};

                        for (String extension : okFileExtensions) {

                            Log.d("ChatMessagesAdapter", file.getName().toLowerCase());
                            Log.d("ChatMessagesAdapter", extension);
                            if (file.getName().toLowerCase().endsWith(extension)) {
                                Log.d("ChatMessagesAdapter", "true");

                                if (BitmapManager.checkValidImage(file)) {
                                    BitmapManager.loadBitmap(file, file.getPath().hashCode(), holder.imageMessage);
                                    holder.imageMessage.setVisibility(View.VISIBLE);
                                    holder.imageMessageFrame.setVisibility(View.VISIBLE);
                                    holder.imageMessage.setOnClickListener(new View.OnClickListener() {
                                        public void onClick(View v) {
                                            Intent i = new Intent();
                                            i.setAction(android.content.Intent.ACTION_VIEW);
                                            i.setDataAndType(Uri.fromFile(file), "image/*");
                                            ChatMessagesAdapter.this.context.startActivity(i);

                                        }
                                    });
                                    if (msg.received) {
                                        holder.message.setVisibility(View.GONE);
                                        holder.title.setVisibility(View.GONE);
                                        holder.progressText.setVisibility(View.GONE);
                                    } else {
                                        holder.padding.setVisibility(View.VISIBLE);
                                    }
                                }

                                break; // break for loop
                            }

                        }
                    }
                }

                break;

            case Constants.MESSAGE_TYPE_ACTION:

                holder.time.setGravity(Gravity.CENTER);
                holder.layout.setGravity(Gravity.CENTER);
                holder.message.setTextColor(context.getResources().getColor(R.color.gray_darker));
                holder.row.setGravity(Gravity.CENTER);
                holder.background.setBackgroundColor(context.getResources().getColor(R.color.white_absolute));
                holder.message.setTextSize(12);

                break;
        }

        holder.time.setText(PrettyTimestamp.prettyTimestamp(msg.time, true));
        holder.time.setVisibility(View.VISIBLE);

        if (!animatedIds.contains(id)) {
            holder.row.startAnimation(anim);
            animatedIds.add(id);
        }

        holder.row.setOnLongClickListener(new View.OnLongClickListener() {
            @Override
            public boolean onLongClick(View view) {
                if (type == Constants.MESSAGE_TYPE_OWN || type == Constants.MESSAGE_TYPE_FRIEND) {
                    final AlertDialog.Builder builder = new AlertDialog.Builder(context);
                    CharSequence[] items = new CharSequence[]{
                            context.getResources().getString(R.string.message_copy),
                            context.getResources().getString(R.string.message_delete)
                    };
                    builder.setCancelable(true)
                            .setItems(items, new DialogInterface.OnClickListener() {
                                public void onClick(DialogInterface dialog, int index) {
                                    switch (index) {
                                        case 0: //Copy
                                            ClipboardManager clipboard = (ClipboardManager) context.getSystemService(context.CLIPBOARD_SERVICE);
                                            clipboard.setText(m);
                                            break;
                                        case 1: //Delete
                                            class DeleteMessage extends AsyncTask<Void, Void, Void> {
                                                @Override
                                                protected Void doInBackground(Void... params) {
                                                    AntoxDB antoxDB = new AntoxDB(context.getApplicationContext());
                                                    antoxDB.deleteMessage(id);
                                                    antoxDB.close();
                                                    return null;
                                                }

                                                @Override
                                                protected void onPostExecute(Void result) {
                                                    toxSingleton.updateMessages(context);
                                                }

                                            }
                                            new DeleteMessage().execute();
                                            break;
                                    }
                                }
                            });
                    AlertDialog alert = builder.create();
                    alert.show();
                } else {
                    final AlertDialog.Builder builder = new AlertDialog.Builder(context);
                    CharSequence[] items = new CharSequence[]{
                            context.getResources().getString(R.string.message_delete)
                    };
                    builder.setCancelable(true)
                            .setItems(items, new DialogInterface.OnClickListener() {
                                public void onClick(DialogInterface dialog, int index) {
                                    switch (index) {
                                        case 0: //Delete
                                            class DeleteMessage extends AsyncTask<Void, Void, Void> {
                                                @Override
                                                protected Void doInBackground(Void... params) {
                                                    AntoxDB antoxDB = new AntoxDB(context.getApplicationContext());
                                                    antoxDB.deleteMessage(id);
                                                    antoxDB.close();
                                                    return null;
                                                }

                                                @Override
                                                protected void onPostExecute(Void result) {
                                                    toxSingleton.updateMessages(context);
                                                }

                                            }
                                            new DeleteMessage().execute();

                                            break;
                                    }
                                }
                            });
                    AlertDialog alert = builder.create();
                    alert.show();
                }
                return true;
            }
        });
    }

    @Override
    public int getViewTypeCount() {
        return 4;
    }

    private void ownMessage(ChatMessagesHolder holder) {
        holder.time.setGravity(Gravity.RIGHT);
        holder.layout.setGravity(Gravity.RIGHT);
        holder.message.setTextColor(context.getResources().getColor(R.color.white_absolute));
        holder.row.setGravity(Gravity.RIGHT);
        holder.background.setBackgroundDrawable(context.getResources().getDrawable(R.drawable.chatright));
        holder.background.setPadding(1*density*paddingscale, 1*density*paddingscale, 6*density + 1*density*paddingscale, 1*density*paddingscale);
    }

    private void friendMessage(ChatMessagesHolder holder) {
        holder.message.setTextColor(context.getResources().getColor(R.color.black));
        holder.background.setBackgroundDrawable(context.getResources().getDrawable(R.drawable.chatleft));
        holder.background.setPadding(6 * density + 1 * density * paddingscale, 1 * density * paddingscale, 1 * density * paddingscale, 1 * density * paddingscale);
        holder.time.setGravity(Gravity.LEFT);
        holder.layout.setGravity(Gravity.LEFT);
        holder.row.setGravity(Gravity.LEFT);
    }

    static class ChatMessagesHolder {
        LinearLayout row;
        LinearLayout layout;
        LinearLayout background;
        TextView message;
        TextView time;
        ImageView sent;
        ImageView received;
        ImageView imageMessage;
        FrameLayout imageMessageFrame;
        TextView title;
        ProgressBar progress;
        TextView progressText;
        View padding;
        LinearLayout buttons;
        View accept;
        View reject;
    }

}

File: app/src/main/scala/im/tox/antox/adapters/ChatMessagesAdapter.scala
package im.tox.antox.adapters

import java.io.File
import java.util
import java.util.Random

import android.app.AlertDialog
import android.content.{Context, DialogInterface, Intent}
import android.graphics.{Color, Typeface}
import android.net.Uri
import android.os.{Build, Environment}
import android.text.{ClipboardManager, Html}
import android.view.animation.{Animation, AnimationUtils}
import android.view.{Gravity, LayoutInflater, View, ViewGroup}
import android.widget._
import im.tox.antox.adapters.ChatMessagesAdapter._
import im.tox.antox.data.{AntoxDB, State}
import im.tox.antox.tox.ToxSingleton
import im.tox.antox.utils.{BitmapManager, Constants, TimestampUtils}
import im.tox.antox.wrapper.{Message, MessageType}
import im.tox.antoxnightly.R
import rx.lang.scala.Observable
import rx.lang.scala.schedulers.IOScheduler


object ChatMessagesAdapter {

  class ChatMessagesHolder {

    var row: LinearLayout = _

    var layout: LinearLayout = _

    var background: LinearLayout = _

    var message: TextView = _

    var time: TextView = _

    var imageMessage: ImageView = _

    var imageMessageFrame: FrameLayout = _

    var title: TextView = _

    var progress: ProgressBar = _

    var progressText: TextView = _

    var padding: View = _

    var buttons: LinearLayout = _

    var accept: View = _

    var reject: View = _

    var bubble: LinearLayout = _

    var wrapper: LinearLayout = _

    var sentTriangle: View = _

    var receivedTriangle: View = _
  }
}

class ChatMessagesAdapter(var context: Context, messages: util.ArrayList[Message], ids: util.HashSet[Integer])
  extends ArrayAdapter[Message](context, R.layout.chat_message_row, messages) {

  var layoutResourceId: Int = R.layout.chat_message_row

  private val density: Int = context.getResources.getDisplayMetrics.density.toInt

  private val anim: Animation = AnimationUtils.loadAnimation(this.context, R.anim.abc_slide_in_bottom)

  private val mInflater: LayoutInflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE).asInstanceOf[LayoutInflater]

  private val animatedIds: util.HashSet[Integer] = ids

  override def getView(position: Int, convertView: View, parent: ViewGroup): View = {
    var view: View = null
    var holder: ChatMessagesHolder = null

    val msg = getItem(position)

    //FIXME
    val lastMsg: Message = null
    val nextMsg: Message = null

    if (convertView == null) {
      view = mInflater.inflate(this.layoutResourceId, parent, false)

      holder = new ChatMessagesHolder()
      holder.message = view.findViewById(R.id.message_text).asInstanceOf[TextView]
      holder.layout = view.findViewById(R.id.message_text_layout).asInstanceOf[LinearLayout]
      holder.row = view.findViewById(R.id.message_row_layout).asInstanceOf[LinearLayout]
      holder.background = view.findViewById(R.id.message_text_background).asInstanceOf[LinearLayout]
      holder.time = view.findViewById(R.id.message_text_date).asInstanceOf[TextView]
      holder.title = view.findViewById(R.id.message_title).asInstanceOf[TextView]
      holder.progress = view.findViewById(R.id.file_transfer_progress).asInstanceOf[ProgressBar]
      holder.imageMessage = view.findViewById(R.id.message_sent_photo).asInstanceOf[ImageView]
      holder.imageMessageFrame = view.findViewById(R.id.message_sent_photo_frame).asInstanceOf[FrameLayout]
      holder.progressText = view.findViewById(R.id.file_transfer_progress_text).asInstanceOf[TextView]
      holder.padding = view.findViewById(R.id.file_transfer_padding)
      holder.buttons = view.findViewById(R.id.file_buttons).asInstanceOf[LinearLayout]
      holder.accept = view.findViewById(R.id.file_accept_button)
      holder.reject = view.findViewById(R.id.file_reject_button)
      holder.bubble = view.findViewById(R.id.message_bubble).asInstanceOf[LinearLayout]
      holder.wrapper = view.findViewById(R.id.message_background_wrapper).asInstanceOf[LinearLayout]
      holder.sentTriangle = view.findViewById(R.id.sent_triangle)
      holder.receivedTriangle = view.findViewById(R.id.received_triangle)

      view.setTag(holder)
    } else {
      view = convertView
      holder = view.getTag.asInstanceOf[ChatMessagesHolder]
    }

    /* if (cursor.moveToPrevious()) {
      lastMsg = getItem(position - 1)
    }
    cursor.moveToNext()

    if (cursor.moveToNext()) {
      nextMsg = chatMessageFromCursor(cursor)
    }
    cursor.moveToPrevious() */

    holder.message.setTextSize(16)
    holder.message.setVisibility(View.GONE)
    holder.time.setVisibility(View.GONE)
    holder.title.setVisibility(View.GONE)
    holder.progress.setVisibility(View.GONE)
    holder.imageMessage.setVisibility(View.GONE)
    holder.imageMessageFrame.setVisibility(View.GONE)
    holder.progressText.setVisibility(View.GONE)
    holder.padding.setVisibility(View.GONE)
    holder.buttons.setVisibility(View.GONE)
    holder.sentTriangle.setVisibility(View.GONE)
    holder.receivedTriangle.setVisibility(View.GONE)
    setAlpha(holder.bubble, 1f)
    msg.`type` match {
      case MessageType.OWN | MessageType.GROUP_OWN =>
        holder.message.setText(msg.message)
        ownMessage(holder)
        holder.message.setVisibility(View.VISIBLE)
        if (!msg.received) {
          setAlpha(holder.bubble, 0.5f)
        }

      case MessageType.GROUP_PEER =>
        holder.message.setText(msg.message)
        holder.title.setText(msg.sender_name)
        groupMessage(holder, genNameColor(msg.sender_name))
        holder.message.setVisibility(View.VISIBLE)
        if (lastMsg == null || msg.sender_name != lastMsg.sender_name) {
          holder.title.setVisibility(View.VISIBLE)
        }
        if (!msg.received) {
          setAlpha(holder.bubble, 0.5f)
        }

      case MessageType.FRIEND =>
        holder.message.setText(msg.message)
        contactMessage(holder)
        holder.message.setVisibility(View.VISIBLE)

      case MessageType.FILE_TRANSFER | MessageType.FILE_TRANSFER_FRIEND =>
        if (msg.`type` == MessageType.FILE_TRANSFER) {
          ownMessage(holder)
          val split = msg.message.split("/")
          holder.message.setText(split(split.length - 1))
          holder.message.setVisibility(View.VISIBLE)
        } else {
          contactMessage(holder)
          holder.message.setText(msg.message)
          holder.message.setVisibility(View.VISIBLE)
        }
        holder.title.setVisibility(View.VISIBLE)
        holder.title.setText(R.string.chat_file_transfer)
        if (msg.received) {
          holder.progressText.setText(context.getResources.getString(R.string.file_finished))
          holder.progressText.setVisibility(View.VISIBLE)
        } else {
          if (msg.sent) {
            if (msg.message_id != -1) {
              holder.progress.setVisibility(View.VISIBLE)
              holder.progress.setMax(msg.size)
              holder.progress.setProgress(State.transfers.getProgress(msg.id).toInt)
              val mProgress = State.transfers.getProgressSinceXAgo(msg.id, 500)
              val bytesPerSecond = mProgress match {
                case Some(p) => ((p._1 * 1000) / p._2).toInt
                case None => 0
              }
              if (bytesPerSecond != 0) {
                val secondsToComplete = msg.size / bytesPerSecond
                holder.progressText.setText(java.lang.Integer.toString(bytesPerSecond / 1024) + " KiB/s, " +
                  context.getResources.getString(R.string.file_time_remaining, secondsToComplete.toString))
              } else {
                holder.progressText.setText(java.lang.Integer.toString(bytesPerSecond / 1024) + " KiB/s")
              }
              holder.progressText.setVisibility(View.VISIBLE)
            } else {
              //FIXME this should be "Failed" - fix the DB bug
              holder.progressText.setText(context.getResources.getString(R.string.file_finished))
              holder.progressText.setVisibility(View.VISIBLE)
            }
          } else {
            if (msg.message_id != -1) {
              if (msg.isMine) {
                holder.progressText.setText(context.getResources.getString(R.string.file_request_sent))
              } else {
                holder.progressText.setText("")
                holder.buttons.setVisibility(View.VISIBLE)
                holder.accept.setOnClickListener(new View.OnClickListener() {

                  override def onClick(view: View) {
                    State.transfers.acceptFile(msg.key, msg.message_id, context)
                  }
                })
                holder.reject.setOnClickListener(new View.OnClickListener() {

                  override def onClick(view: View) {
                    State.transfers.rejectFile(msg.key, msg.message_id, context)
                  }
                })
              }
            } else {
              holder.progressText.setText(context.getResources.getString(R.string.file_rejected))
            }
            holder.progressText.setVisibility(View.VISIBLE)
          }
        }
        if (msg.received || msg.isMine) {
          var f: File = null
          if (msg.message.contains("/")) {
            f = new File(msg.message)
          } else {
            f = new File(Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_DOWNLOADS),
              Constants.DOWNLOAD_DIRECTORY)
            f = new File(f.getAbsolutePath + "/" + msg.message)
          }
          if (f.exists() && (msg.received || msg.isMine)) {
            val file = f
            val okFileExtensions = Array("jpg", "png", "gif", "jpeg")
            for (extension <- okFileExtensions) {
              //Log.d("ChatMessagesAdapter", file.getName.toLowerCase())
              //Log.d("ChatMessagesAdapter", extension)
              if (file.getName.toLowerCase.endsWith(extension)) {
                // Set a placeholder in the image in case bitmap needs to be loaded from disk
                if (msg.isMine)
                  holder.imageMessage.setImageResource(R.drawable.sent)
                else
                  holder.imageMessage.setImageResource(R.drawable.received)

                BitmapManager.load(file, holder.imageMessage, isAvatar = false)
                holder.imageMessage.setVisibility(View.VISIBLE)
                holder.imageMessageFrame.setVisibility(View.VISIBLE)
                holder.imageMessage.setOnClickListener(new View.OnClickListener() {
                  def onClick(v: View) {
                    val i = new Intent()
                    i.setAction(android.content.Intent.ACTION_VIEW)
                    i.setDataAndType(Uri.fromFile(file), "image/*")
                    ChatMessagesAdapter.this.context.startActivity(i)
                  }
                })
                holder.message.setVisibility(View.GONE)
                holder.title.setVisibility(View.GONE)
                holder.progressText.setVisibility(View.GONE)
              }
              //break
            }
          }
        }

      case MessageType.ACTION =>
        actionMessage(holder)
        holder.message.setText(Html.fromHtml("<b>" + msg.sender_name + "</b> ") + msg.message)

    }

    //TODO: Only show a timestamp if the next message is more than a minute after this one
    if (nextMsg == null ||
      (nextMsg == null && lastMsg == null) ||
      msg.sender_name != nextMsg.sender_name) {
      holder.time.setText(TimestampUtils.prettyTimestamp(msg.timestamp, isChat = true))
      holder.time.setVisibility(View.VISIBLE)
    } else {
      holder.time.setVisibility(View.GONE)
    }

    if (!animatedIds.contains(msg.id)) {
      holder.row.startAnimation(anim)
      animatedIds.add(msg.id)
    }
    view.setOnLongClickListener(new View.OnLongClickListener() {

      override def onLongClick(view: View): Boolean = {
        if (msg.`type` == MessageType.OWN || msg.`type` == MessageType.FRIEND) {
          val builder = new AlertDialog.Builder(context)
          val items = Array[CharSequence](context.getResources.getString(R.string.message_copy), context.getResources.getString(R.string.message_delete))
          builder.setCancelable(true).setItems(items, new DialogInterface.OnClickListener() {

            def onClick(dialog: DialogInterface, index: Int) = index match {
              case 0 =>
                val clipboard = context.getSystemService(Context.CLIPBOARD_SERVICE).asInstanceOf[ClipboardManager]
                clipboard.setText(msg.message)

              case 1 =>
                Observable[Boolean](subscriber => {
                  val antoxDB = new AntoxDB(context.getApplicationContext)
                  antoxDB.deleteMessage(msg.id)
                  antoxDB.close()
                  ToxSingleton.updateMessages(context)
                  subscriber.onCompleted()
                }).subscribeOn(IOScheduler()).subscribe()

            }
          })
          val alert = builder.create()
          alert.show()
        } else {
          val builder = new AlertDialog.Builder(context)
          val items = Array[CharSequence](context.getResources.getString(R.string.message_delete))
          builder.setCancelable(true).setItems(items, new DialogInterface.OnClickListener() {

            def onClick(dialog: DialogInterface, index: Int) = index match {
              case 0 =>
                Observable[Boolean](subscriber => {
                  val antoxDB = new AntoxDB(context.getApplicationContext)
                  antoxDB.deleteMessage(msg.id)
                  antoxDB.close()
                  ToxSingleton.updateMessages(context)
                  subscriber.onCompleted()
                }).subscribeOn(IOScheduler()).subscribe()

            }
          })
          val alert = builder.create()
          alert.show()
        }
        true
      }
    })

    view
  }

  override def getViewTypeCount: Int = MessageType.values.size

  //override def newDropDownView(context: Context, cursor: Cursor, parent: ViewGroup): View = super.newDropDownView(context, cursor, parent)

  //utility method to set view's alpha on honeycomb+ devices,
  //does nothing on pre-honeycomb devices because setAlpha is unsupported
  private def setAlpha(view: View, value: Float): Unit = {
    if (Build.VERSION.SDK_INT < Build.VERSION_CODES.HONEYCOMB) {
      //do nothing
    } else {
      view.setAlpha(value)
    }
  }

  private def shouldGreentext(message: String): Boolean = {
    message.startsWith(">")
  }

  private def genNameColor(name: String): Int = {
    val goldenRatio = 0.618033988749895
    val hue: Double = (new Random(name.hashCode).nextFloat() + goldenRatio) % 1
    Color.HSVToColor(Array(hue.asInstanceOf[Float] * 360, 0.5f, 0.7f))
  }

  private def ownMessage(holder: ChatMessagesHolder) {
    holder.time.setGravity(Gravity.RIGHT)
    holder.layout.setGravity(Gravity.RIGHT)
    holder.sentTriangle.setVisibility(View.VISIBLE)
    if (shouldGreentext(holder.message.getText.toString)) {
      holder.message.setTextColor(context.getResources.getColor(R.color.green_light))
    } else {
      holder.message.setTextColor(context.getResources.getColor(R.color.white))
    }

    holder.row.setGravity(Gravity.RIGHT)
    holder.wrapper.setGravity(Gravity.RIGHT)
    holder.background.setBackgroundDrawable(context.getResources.getDrawable(R.drawable.conversation_item_sent_shape))
    holder.background.setPadding(8 * density, 8 * density, 8 * density, 8 * density)
  }

  private def groupMessage(holder: ChatMessagesHolder, nameColor: Int) {
    contactMessage(holder)
    holder.title.setTextColor(nameColor)
    holder.title.setTypeface(holder.title.getTypeface, Typeface.BOLD)
  }

  private def contactMessage(holder: ChatMessagesHolder) {
    if (shouldGreentext(holder.message.getText.toString)) {
      holder.message.setTextColor(context.getResources.getColor(R.color.green))
    } else {
      holder.message.setTextColor(context.getResources.getColor(R.color.black))
    }
    holder.receivedTriangle.setVisibility(View.VISIBLE)
    holder.time.setGravity(Gravity.LEFT)
    holder.layout.setGravity(Gravity.LEFT)
    holder.row.setGravity(Gravity.LEFT)
    holder.wrapper.setGravity(Gravity.LEFT)
    holder.background.setBackgroundDrawable(context.getResources.getDrawable(R.drawable.conversation_item_received_shape))
    holder.background.setPadding(8 * density, 8 * density, 8 * density, 8 * density)
  }

  private def actionMessage(holder: ChatMessagesHolder) {
    holder.message.setVisibility(View.VISIBLE)
    holder.message.setTextSize(18)
    holder.message.setTextColor(context.getResources.getColor(R.color.black))
    holder.time.setGravity(Gravity.CENTER)
    holder.layout.setGravity(Gravity.CENTER)
    holder.row.setGravity(Gravity.CENTER)
    holder.wrapper.setGravity(Gravity.CENTER)
    holder.background.setBackgroundColor(context.getResources.getColor(R.color.white_absolute))
    holder.background.setPadding(8 * density, 8 * density, 8 * density, 8 * density)
    holder.time.setVisibility(View.VISIBLE)
  }
}
